BridgeCareApp/BridgeCare/DataAccessLayer/AttributeByYear.cs

```diff
@@ -0,0 +1,117 @@
+                List<AttributeByYearModel> returnvals = DataTableToAttributeList(dt);
+                return returnvals;
+            }
+            catch (SqlException ex)
+            {
+                HandleException.SqlError(ex, "Query:" + select);
+            }
+            catch (OutOfMemoryException ex)
+            {
+                HandleException.OutOfMemoryError(ex);
+            }
+
+            return null;
+        }
+
+        public List<AttributeByYearModel> GetHistoricalAttributes(int NetworkId, int sectionID, BridgeCareContext db)
+        {
+            DataTable dt = null;
+            var select = String.Format(
+                "SELECT * FROM SEGMENT_{0}_NS0 WHERE SectionID={1}",
+                NetworkId, sectionID);
+
+            try
+            {
+                dt = UtilityFunctions.NonEntitySQLQuery(select,db);
+                List<AttributeByYearModel> returnvals = DataTableToAttributeList(dt);
+                return returnvals;
+            }
+            catch (SqlException ex)
+            {
+                HandleException.SqlError(ex, "Query:" + select);
+            }
+            catch (OutOfMemoryException ex)
+            {
+                HandleException.OutOfMemoryError(ex);
+            }
+            return null;
+        }
+
+        public List<AttributeByYearModel> DataTableToAttributeList(DataTable dt)
+        {
+            List<AttributeByYearModel> returnValues = new List<AttributeByYearModel>();
+
+            foreach (DataColumn col in dt.Columns)
+            {
+                string value = dt.Rows[0][col.ColumnName].ToString();
+
+                string[] tokens = col.ColumnName.Split('_');
+                if (UtilityFunctions.IsIamYear(tokens.Last()) && value.Length > 0)
+                {
+                    //yes means this is an attribute_year column, subtract 5 digits, 4 for year and one '_'
+
+                    string name = col.ColumnName.Substring(0, col.ColumnName.Length - 5);
+                    AttributeYearlyValueModel yvm = new AttributeYearlyValueModel();
+
+                    yvm.year = Convert.ToInt32(tokens.Last());
+                    yvm.value = value;
+
+                    AttributeByYearModel temp = IndexofName(returnValues, name);
+                    if (temp == null)
+                    {
+                        temp = new AttributeByYearModel();
+                        temp.name = name;
+                    }
+
+                    temp.yearlyvalues.Add(yvm);
+                    returnValues.Add(temp);
+                }
+
+
+            }
+            return returnValues;
+        }
+        // return reference to array element where name  matches findname
+        // encapsulated here as brute force version , could utilize map instead
+        public AttributeByYearModel IndexofName(List<AttributeByYearModel> lam, string findname)
+        {
+            foreach (AttributeByYearModel aym in lam)
```

Comment:
Use a LINQ expression here.

`return lam.Find(_ => _.name == aym.name)`

If you need to perform a similar search on IEnumerable, use the .Where.FirstOrDefault() methods.

Note that if no match is found it will return the default value for a reference type (null). Again not desirable to return null from your APIs. See if there is a default value that can be returned and use it.

Reply:
That line alone is the same as the entire function, got rid of function